### PR TITLE
Binary cross entropy logits loss

### PR DESCRIPTION
**Pull request by mayankray2020**
Addition of Binary cross entropy logits loss in loss_functions

### CODE REVIEW

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+void BCELossWithLogits<InputDataType, OutputDataType>::Backward(
+    const PredictionType& prediction,
+    const TargetType& target,
+    LossType& loss)
+{
+  typedef typename PredictionType::elem_type ElemType;
+
+  ElemType sigmoid = 1 / (1 + arma::exp(-prediction));
+
+  loss = ((1. - target) / (1. - prediction + eps) - target / (prediction + eps)) % sigmoid % (1 - sigmoid);
```
Comment (abh2k) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:56]:
```suggestion
  loss = ((1. - target) / (1. - sigmoid + eps) - target / (sigmoid + eps)) % sigmoid % (1 - sigmoid);
```

File: src/mlpack/tests/loss_functions_test.cpp [hunk 2]
```diff
@@ -320,6 +321,48 @@ TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
...
+  target1 = arma::zeros(1, 8);
+  double error1 = module1.Forward(input1, target1);
+  REQUIRE(error1 - 7.792616 == Approx(0.0).margin(2e-5));
+
+  input2 = arma::mat("1.5 1.5 1.5 1.5 1.5");
+  target2 = arma::ones(1, 5);
+  input2.reshape(2, 3);
+  target2.reshape(2, 3);
+  double error2 = module2.Forward(input2, target2);
+  double loss2 = -std::log(1 / (1 + std::exp(1.5)));
```
Comment (abh2k) [inline src/mlpack/tests/loss_functions_test.cpp:344]:
```suggestion
  double loss2 = std::log(1 + std::exp(-1.5));
```

File: src/mlpack/tests/loss_functions_test.cpp [hunk 3]
```diff
@@ -320,6 +321,48 @@ TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
...
+{
+  arma::mat input1, input2, input3, output, target1, target2, target3;
+  BCELossWithLogits<> module1(1e-6, false);
+  BCELossWithLogits<> module2(1e-6, true);
+  // Test the Forward function on a user generator input and compare it against
+  // the manually calculated result.
+  input1 = arma::mat("0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5");
+  target1 = arma::zeros(1, 8);
+  double error1 = module1.Forward(input1, target1);
+  REQUIRE(error1 - 7.792616 == Approx(0.0).margin(2e-5));
```
Comment (abh2k) [inline src/mlpack/tests/loss_functions_test.cpp:337]:
```suggestion
  REQUIRE(error1 - 3.792615 == Approx(0.0).margin(2e-5));
```

File: src/mlpack/tests/loss_functions_test.cpp [hunk 4]
```diff
@@ -320,6 +321,48 @@ TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
...
+  target2.reshape(2, 3);
+  double error2 = module2.Forward(input2, target2);
+  double loss2 = -std::log(1 / (1 + std::exp(1.5)));
+  REQUIRE((error2 - loss2) == Approx(0.0).margin(2e-5));
+
+  // Test the Backward function.
+  module1.Backward(input1, target1, output);
+  for (double el : output)
+  {
+    REQUIRE(el - 0.077807 == Approx(0.0).margin(5e-6));
```
Comment (abh2k) [inline src/mlpack/tests/loss_functions_test.cpp:351]:
```suggestion
    REQUIRE(el - (-0.377540) == Approx(0.0).margin(5e-6));
```

**Comment by abh2k** [comment]
The test cases are here https://colab.research.google.com/drive/1imsjBEUxd_F97z91Jihx4nS3iVmI58OV?usp=sharing

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename PredictionType, typename TargetType, typename LossType>
+void BCELossWithLogits<InputDataType, OutputDataType>::Backward(
+    const PredictionType& prediction,
+    const TargetType& target,
+    LossType& loss)
+{
+  typedef typename PredictionType::elem_type ElemType;
```
Comment (abh2k) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:52]:
```suggestion
```

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  ElemType sigmoid = 1 / (1 + arma::exp(-prediction));
```
Comment (abh2k) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:54]:
```suggestion
  LossType sigmoid = 1 / (1 + arma::exp(-prediction));
```

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+template<typename InputDataType, typename OutputDataType>
+template<typename PredictionType, typename TargetType>
+typename PredictionType::elem_type
+BCELossWithLogits<InputDataType, OutputDataType>::Forward(
+    const PredictionType& prediction,
+    const TargetType& target)
+{
+  typedef typename PredictionType::elem_type ElemType;
+
+  ElemType sigmoid = 1 / (1 + arma::exp(-prediction));
```
Comment (abh2k) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:37]:
```suggestion
  ElemType sigmoid = arma::pow(1 + arma::exp(-prediction), -1);
```

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+template<typename InputDataType, typename OutputDataType>
+template<typename PredictionType, typename TargetType, typename LossType>
+void BCELossWithLogits<InputDataType, OutputDataType>::Backward(
+    const PredictionType& prediction,
+    const TargetType& target,
+    LossType& loss)
+{
+
+  LossType sigmoid = 1 / (1 + arma::exp(-prediction));
```
Comment (abh2k) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:53]:
```suggestion
  LossType sigmoid = arma::pow(1 + arma::exp(-prediction), -1);
```

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+template<typename InputDataType, typename OutputDataType>
+template<typename PredictionType, typename TargetType>
+typename PredictionType::elem_type
+BCELossWithLogits<InputDataType, OutputDataType>::Forward(
+    const PredictionType& prediction,
+    const TargetType& target)
+{
+  typedef typename PredictionType::elem_type ElemType;
+
+  ElemType sigmoid = arma::pow(1 + arma::exp(-prediction), -1);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:37]: The output is not a scalar (`ElemType`) but a matrix, so you have to store the result as `PredictionType` or `OutputDataType`.

Comment (abh2k) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:37]:
```suggestion

  OutputDataType sigmoid = arma::pow(1 + arma::exp(-prediction), -1);
```

File: src/mlpack/tests/loss_functions_test.cpp [hunk 10]
```diff
@@ -320,6 +321,48 @@ TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
...
+ */
+TEST_CASE("SimpleBinaryCrossEntropyWithLogitsLossTest", "[LossFunctionsTest]")
+{
+  arma::mat input1, input2, input3, output, target1, target2, target3;
+  BCELossWithLogits<> module1(1e-6, false);
+  BCELossWithLogits<> module2(1e-6, true);
+  // Test the Forward function on a user generator input and compare it against
+  // the manually calculated result.
+  input1 = arma::mat("0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5");
+  target1 = arma::zeros(1, 8);
```
Comment (abh2k) [inline src/mlpack/tests/loss_functions_test.cpp:335]:
```suggestion
  target1 = arma::ones(1, 8);
```

I checked the test cases using `ones` .

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    loss /= prediction.n_elem;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename Archive>
+void BCELossWithLogits<InputDataType, OutputDataType>::serialize(
+    Archive& ar,
+    const uint32_t /* version */)
+{
+  ar(CEREAL_NVP(eps));
```
Comment (Abilityguy) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:66]:
```suggestion
  ar(CEREAL_NVP(eps));
  ar(CEREAL_NVP(reduction));
```
I believe we should store the reduction parameter too.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 12]
```diff
@@ -320,6 +321,48 @@ TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
...
+  BCELossWithLogits<> module2(1e-6, true);
+  // Test the Forward function on a user generator input and compare it against
+  // the manually calculated result.
+  input1 = arma::mat("0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5");
+  target1 = arma::ones(1, 8);
+  double error1 = module1.Forward(input1, target1);
+  REQUIRE(error1 - 3.792615 == Approx(0.0).margin(2e-5));
+
+  input2 = arma::mat("1.5 1.5 1.5 1.5 1.5");
+  target2 = arma::ones(1, 5);
```
Comment (abh2k) [inline src/mlpack/tests/loss_functions_test.cpp:340]:
```suggestion
  input2 = arma::mat("1.5 1.5 1.5 1.5 1.5 1.5");
  target2 = arma::ones(1, 6);
```

File: src/mlpack/tests/loss_functions_test.cpp [hunk 13]
```diff
@@ -320,6 +321,48 @@ TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
...
+    REQUIRE(el - (-0.377540) == Approx(0.0).margin(5e-6));
+  }
+  REQUIRE(output.n_rows == input1.n_rows);
+  REQUIRE(output.n_cols == input1.n_cols);
+
+  module1.Backward(input2, target2, output);
+  for (double el : output)
+  {
+    REQUIRE(el - -0.036485 == Approx(0.0).margin(5e-6));
+  }
```
Comment (abh2k) [inline src/mlpack/tests/loss_functions_test.cpp:360]:
```suggestion
  module2.Backward(input2, target2, output);
  for (double el : output)
  {
    REQUIRE(el - (-0.030404) == Approx(0.0).margin(5e-6));
  }
```

**Comment by mayankray2020** [comment]
@zoq please review this.

File: HISTORY.md [hunk 14]
```diff
@@ -1,5 +1,7 @@
 ### mlpack ?.?.?
 ###### ????-??-??
+  * Added Binary Cross Entropy Logits Loss
```
Comment (zoq) [inline HISTORY.md:3]: The PR number is missing.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp (hunk 11 above)
Comment (zoq) [reply src/mlpack/methods/ann/loss_functions/binary_cross_entropy_logits_loss_impl.hpp:66]: Agreed.

**Review by zoq — COMMENTED** [review]
This looks good to me, just some minor issue to fix before we can merge this in.

**Review by zoq — APPROVED** [review]
No further comments from my side, thanks for putting this together!

**Comment by rcurtin** [comment]
This one got approved, but never merged---should we reopen and merge it?
